pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+        probs_or_sample_measure = Sample in return_types or Probability in return_types
+        if probs_or_sample_measure and len(set_of_return_types) > 1:
+            raise ValueError(
+                "Using the JAX interface, sample and probability measurements cannot be mixed with other measurement types."
+            )
+
+        if Probability in return_types:
+            set_len_wires = set(len(o.wires) for o in t.observables)
+            if len(set_len_wires) > 1:
+                raise ValueError(
+                    "Using the JAX interface, multiple probability measurements need to have the same number of wires specified."
+                )
+
+
+def _execute(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        new_tapes = []
+
+        for t, a in zip(tapes, params):
+            new_tapes.append(t.copy(copy_operations=True))
+            new_tapes[-1].set_parameters(a)
+
+        with qml.tape.Unwrap(*new_tapes):
+            res, _ = execute_fn(new_tapes, **gradient_kwargs)
+
+        if len(res) > 1:
+            res = [jnp.array(r) for r in res]
+        else:
+            res = jnp.array(res)
+
+        return res
+
+    def wrapped_exec_fwd(params):
+        return wrapped_exec(params), params
+
+    def wrapped_exec_bwd(params, g):
+
+        if isinstance(gradient_fn, qml.gradients.gradient_transform):
+
+            args = tuple(params) + (g,)
+            new_tapes = []
+            p = args[:-1]
+            dy = args[-1]
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+                new_tapes[-1].trainable_params = t.trainable_params
+
+            vjp_tapes, processing_fn = qml.gradients.batch_vjp(
+                new_tapes,
+                dy,
+                gradient_fn,
+                reduction="append",
+                gradient_kwargs=gradient_kwargs,
+            )
+
+            partial_res = execute_fn(vjp_tapes)[0]
+            res = processing_fn(partial_res)
+            vjps = jnp.concatenate(res)
+
+            param_idx = 0
+            res = []
+
+            # Group the vjps based on the parameters of the tapes
+            for p in params:
+                param_vjp = vjps[param_idx : param_idx + len(p)]
+                res.append(param_vjp)
+                param_idx += len(p)
+
+            # Unwrap partial results into ndim=0 arrays to allow
+            # differentiability with JAX
+            # E.g.,
+            # [DeviceArray([-0.9553365], dtype=float32), DeviceArray([0., 0.],
+            # dtype=float32)]
+            # is mapped to
+            # [[DeviceArray(-0.9553365, dtype=float32)], [DeviceArray(0.,
+            # dtype=float32), DeviceArray(0., dtype=float32)]].
```

Comment:
This looks kinda like `qml.math.unstack`?

Reply:
Oh wait, it looks like `unstack` is not properly defined for JAX :(

![image](https://user-images.githubusercontent.com/2959003/150503106-79dec66e-bcc0-4e8e-b34c-ce80fddcd3d8.png)

The output is a `list[list[float]]` rather than `list[list[DeviceArray]]`.

Reply:
Hmm, good point! Maybe it could be a next step to:
1. Fix `qml.math.unstack`;
2. Swap the logic to use that here.

Reply:
Weird, it seems to work for me both in a notebook and in the terminal :thinking: 

```python
import pennylane as qml
import jax
from jax import numpy as jnp

x = [
    jnp.array([-0.9553365], dtype=jnp.float32),
    jnp.array([0., 0.], dtype=jnp.float32)
]
[qml.math.unstack(a) for a in x]
```
```
[[DeviceArray(-0.9553365, dtype=float32)],
 [DeviceArray(0., dtype=float32), DeviceArray(0., dtype=float32)]]
```
Switching to this.

Reply:
that's odd!